### PR TITLE
Fix CREATE query constraints

### PR DESCRIPTION
**Pull request by evillique**
### Changelog category (leave one):
- Bug Fix (user-visible misbehavior in official stable or prestable release)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fix CREATE USER/ROLE query settings constraints.

### CODE REVIEW

File: src/Interpreters/Access/InterpreterCreateSettingsProfileQuery.cpp [hunk 1]
```diff
@@ -48,16 +48,21 @@ BlockIO InterpreterCreateSettingsProfileQuery::execute()
     else
         getContext()->checkAccess(AccessType::CREATE_SETTINGS_PROFILE);
 
+    std::optional<SettingsProfileElements> settings_from_query;
+    if (query.settings)
+    {
+        settings_from_query = SettingsProfileElements{*query.settings, access_control};
+
+        if (!query.attach)
```
Comment (serxa) [inline src/Interpreters/Access/InterpreterCreateSettingsProfileQuery.cpp:56]: I guess it can be checked earlier

**Review by serxa — COMMENTED** [review]
This change is actually backward incompatible. I'm not sure if we can just say this is a bugfix. Maybe a config option required?

File: src/Interpreters/Access/InterpreterCreateSettingsProfileQuery.cpp (hunk 1 above)
Comment (evillique) [reply src/Interpreters/Access/InterpreterCreateSettingsProfileQuery.cpp:56]: We actually use `settings_from_query` further in the code regardless of `query.attach`, so it seems like checking it earlier would not make sense

**Review by serxa — APPROVED** [review]
Okay, let's merge it as is and maybe, if necessary, add option later

File: tests/integration/test_create_query_constraints/test.py [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+    try:
+        cluster.start()
+        yield cluster
+    finally:
+        cluster.shutdown()
+
+
+def test_create_query_constraints():
+
+    expected_error = "Setting max_threads should not be changed"
```
Comment (vitlibar) [inline tests/integration/test_create_query_constraints/test.py:24]: Is it only the readonly constraint checked from a parent user?
I mean if the parent user (i.e. the `default` user) has a minmax constraint, and then you create a new user with minmax constraint again then your new constraints must be inside the parent user's minmax range.

**Comment by vitlibar** [comment]
> This change is actually backward incompatible. I'm not sure if we can just say this is a bugfix. Maybe a config option required?

It seems to be ok with backward incompatibility without adding an extra setting because we're kind of closing a hole in security here. But for me this PR is rather an improvement, and not a bugfix.

**Review by vitlibar — APPROVED** [review]

File: src/Interpreters/Access/InterpreterCreateUserQuery.cpp [hunk 3]
```diff
@@ -118,12 +118,19 @@ BlockIO InterpreterCreateUserQuery::execute()
...
-    if (!query.cluster.empty())
-        return executeDDLQueryOnCluster(query_ptr, getContext());
+
     std::optional<SettingsProfileElements> settings_from_query;
     if (query.settings)
+    {
         settings_from_query = SettingsProfileElements{*query.settings, access_control};
 
+        if (!query.attach)
+            getContext()->checkSettingsConstraints(*settings_from_query);
```
Comment (vitlibar) [inline src/Interpreters/Access/InterpreterCreateUserQuery.cpp:128]: This is good, but I think for `CREATE / ALTER USER` we also need to add all the current constraints silently in the code to the statement unless they're overridden explicitly by stricter constraints.

For example after executing

```
CREATE USER vasya;
```

the user `vasya` should have the current constraints (not no constraints).

Comment (vitlibar) [reply src/Interpreters/Access/InterpreterCreateUserQuery.cpp:128]: Yes, you're right. But we can silently inherit the current profiles too. Maybe not profiles as a whole, but only their constraints. Something like this:

```
CREATE USER vasya SETTINGS x=8, CONSTRAINTS FROM profile1, CONSTRAINTS FROM profile2, y=9 MIN 5 MAX 10
```

**Comment by vitlibar** [comment]
I've had an idea about an alternative solution. We can switch to a permissive model instead of our current restrictive one. Like it's implemented for access rights.

The idea is the following:
1. by default `CREATE USER` creates a user without any grants. To add grants for a new user you should do that explictly by writing `GRANT TO <username`. So we can make `CREATE USER` without explicitly specified constraints to create a user which cannot change any settings (they all will be readonly for the new user).

2. For any modifications of the constrains via ALTER USER/PROFILE/ROLE we check if that modification is allowed like it's already implemented in this PR. Also we should check and don't allow removing any constraints the current user doesn't have. Inheriting a profile should be allowed only if the current user's constraints allow more.

3. Dropping a profile/role with constraints in this approach won't spoil anything. Because of the permissive model:

```
CREATE USER user1; -- `user1` cannot change any settings
CREATE PROFILE p1 SETTINGS x MIN 0 MAX 10;
CREATE USER user2 SETTINGS PROFILE p1; -- `user2` can change only `x`
DROP PROFILE p1; -- now `user2` cannot change any settings
```

Unfortunately switching to the permissive model is of course backward incompatible.

**Comment by vitlibar** [comment]
> I've had an idea about an alternative solution. We can switch to a permissive model instead of our current restrictive one. Like it's implemented for access rights.
> 
> The idea is the following:
> 
>     1. By default `CREATE USER` creates a user without any grants. To add grants for a new user you should do that explicitly by writing `GRANT TO <username>`. Likewise we can make `CREATE USER` without explicitly specified constraints to create a user which cannot change any settings (they all will be readonly for the new user).
> 
>     2. For any modifications of the constrains via ALTER USER/PROFILE/ROLE we check if that modification is allowed as it's already implemented in this PR. Also we should check and don't allow removing any constraints the current user doesn't have. Inheriting a profile should be allowed only if the current user's constraints are equal or broader.
> 
>     3. Dropping a profile/role with constraints in this approach won't spoil anything. Because of the permissive model:
> 
> 
> ```
> CREATE USER user1; -- `user1` cannot change any settings
> CREATE PROFILE p1 SETTINGS x MIN 0 MAX 10;
> CREATE USER user2 SETTINGS PROFILE p1; -- `user2` can change only `x`
> DROP PROFILE p1; -- now `user2` cannot change any settings
> ```
> 
> Unfortunately switching to the permissive model is of course backward incompatible.

UPD: No, over complicated, don't do that ;)

Let's keep the implementation of this PR as it is now. Because it's obvious and any further constructions are not. Perhaps later we will implement a better solution.

**Comment by serxa** [comment]
Agree. Better solution is needed but it is separate. The only thing that we can improve in this PR is to add checks for ALTERs

File: src/Access/SettingsConstraints.cpp [hunk 4]
```diff
@@ -98,24 +101,20 @@ void SettingsConstraints::check(const Settings & current_settings, const Setting
...
+            old_value = it->second.writability;
 
-            if (new_value != old_value)
-            {
-                if (old_value == SettingConstraintWritability::CONST)
-                    throw Exception("Setting " + element.setting_name + " should not be changed", ErrorCodes::SETTING_CONSTRAINT_VIOLATION);
-
-                if (new_value == SettingConstraintWritability::CHANGEABLE_IN_READONLY &&
-                    old_value == SettingConstraintWritability::WRITABLE)
-                    throw Exception("Setting " + element.setting_name + " should not be changed", ErrorCodes::SETTING_CONSTRAINT_VIOLATION);
```
Comment (serxa) [inline src/Access/SettingsConstraints.cpp]: Denying WRITEABLE -> CHANGABLE_IN_READONLY looked quite right to me (an extension for readonly users), why is this now removed? Maybe we should deny it at least for readonly users?

Comment (evillique) [reply src/Access/SettingsConstraints.cpp]: Do you mean for the users with const/readonly constraints on the setting we're checking? If so, we seem to already do it.

The main reason is that WRITABLE is the default level, and it is WRITABLE regardless of if there was a constraint at all. So with it, it was actually not possible to set CHANGABLE_IN_READONLY at all.

Comment (evillique) [reply src/Access/SettingsConstraints.cpp]: CREATE/ALTER queries are unavailable in `readonly>0`, it is true

**Comment by evillique** [comment]
AST fuzzer (ubsan): https://github.com/ClickHouse/ClickHouse/issues/44313

**Comment by evillique** [comment]
Test failures are unrelated.

**Pull request merged**